src/IROperator.cpp

```diff
@@ -446,9 +447,9 @@ Expr lossless_cast(Type t, Expr e) {
     }
 
     if (const Cast *c = e.as<Cast>()) {
-        if (t.can_represent(c->value.type())) {
-            // We can recurse into widening casts.
-            return lossless_cast(t, c->value);
+        Expr v = lossless_cast(t, c->value);
```

Comment:
I don't think you can recurse into narrowing casts though: consider 
```lossless_cast(UInt16(), cast(UInt8(), cast(UInt16(), e)))```
It would return `e`, which is incorrect.

Reply:
@abadams What is `lossless_cast(UInt(16), cast(Int(8), Variable::make(UInt(16, e)));` supposed to return?

Reply:
In master, it returns `uint16(uint8((uint16)e))`

In my head I was thinking lossless_cast is supposed to be one-to-one, so that means:

if x1 == x2 then lossless_cast(t, x1) == lossless_cast(t, x2)

and

if lossless_cast(t, x1) == lossless_cast(t, x2) then x1 == x2

The new rule violates the first of those two conditions (consider e == 0 vs e == 256), but not the second.

I think the issue here is that this isn't written down very precisely. It just says "provably not losing information". 1-to-1 is more than you need for that condition - it just needs to be invertible, so widening and adding junk to the new high bits would qualify, because you can just strip them again. Really the spec implied by the comment is that not only is it invertible, it's inverted by a cast back to the original type:

cast(x.type(), lossless_cast(t, x)) == x

That spec would permit what you've written. I'm a little nervous about making junk high bits legal though. You'd have to know it was feeding into some instruction that ignored them.

Reply:
@abadams I agree that the change here is incorrect for the reason you mentioned. However, I think the existing code in the master also has a problem. The master returns
`uint16(uint8((uint16)e))` for `lossless_cast(UInt(16), cast(UInt(8), Variable::make(UInt(16, e)));`
but for 
`lossless_cast(UInt(16), cast(**Int(8)**, Variable::make(UInt(16, e)));`  it returns `(uint16)e` which I believe is wrong.

Reply:
Ah, you're right. That's a bug in master!

Reply:
Hey Dillon, I've moved the IROperator changes to separate PR. These changes are however needed for this PR too. Otherwise, find_mpy_ops is not able to find some expressions which we can lossless_cast to the required type.

https://github.com/halide/Halide/pull/5459